libandroid-navigation/src/main/java/com/mapbox/services/android/navigation/v5/route/RouteEngine.java

```diff
@@ -0,0 +1,86 @@
+package com.mapbox.services.android.navigation.v5.route;
+
+import android.support.annotation.NonNull;
+
+import com.mapbox.api.directions.v5.models.DirectionsResponse;
+import com.mapbox.api.directions.v5.models.RouteOptions;
+import com.mapbox.geojson.Point;
+import com.mapbox.services.android.navigation.v5.navigation.NavigationRoute;
+import com.mapbox.services.android.navigation.v5.routeprogress.RouteProgress;
+
+import java.util.ArrayList;
+import java.util.List;
+
+import retrofit2.Call;
+import retrofit2.Callback;
+import retrofit2.Response;
+
+/**
+ * This class can be used to fetch new routes given a {@link Point} origin and
+ * {@link RouteOptions} provided by a {@link RouteProgress}.
+ */
+public class RouteEngine implements Callback<DirectionsResponse> {
+
+  private Callback engineCallback;
+  private RouteProgress routeProgress;
+
+  public RouteEngine(Callback engineCallback) {
+    this.engineCallback = engineCallback;
+  }
+
+  public void fetchRoute(Point origin, RouteProgress routeProgress) {
+    if (routeProgress == null) {
+      return;
+    }
+    this.routeProgress = routeProgress;
+
+    // Calculate remaining waypoints
+    List<Point> coordinates = new ArrayList<>(routeProgress.directionsRoute().routeOptions().coordinates());
```

Comment:
@ericrwolfe starting with this line, does this logic make sense?  I've tested with multi-waypoint / origin + destination routes and is seems to be working fine

Reply:
Hrmm, @bsudekum might be a better person to review this logic. I'm not as versed on how we deal with multiple waypoints.

Reply:
@bsudekum `RouteOptions#coordinates` represents the list of points making up the route.  This would be equivalent to your list of legs with destinations (which we currently don't have on Android)